File: src/main/java/de/mrapp/apriori/ItemSet.java
/*
 * Copyright 2017 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.apriori;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.*;

import static de.mrapp.util.Condition.*;

/**
 * An item set, which may contain several items. An item set may not contain duplicates. The
 * contained items are ordered according to the implementation of their {@link
 * Comparable#compareTo(Object)} method.
 *
 * @param <ItemType> The type of the items, which are contained by the item set
 * @author Michael Rapp
 * @since 1.0.0
 */
public class ItemSet<ItemType> implements SortedSet<ItemType>, Comparable<ItemSet<ItemType>>,
        Serializable, Cloneable {

    /**
     * The constant serial version UID.
     */
    private static final long serialVersionUID = 1L;

    /**
     * A sorted set, which contains the items, which are contained by the item set.
     */
    private final SortedSet<ItemType> items;

    /**
     * The support of the item set.
     */
    private double support;

    /**
     * Creates an empty item set.
     */
    public ItemSet() {
        this.items = new TreeSet<>();
        setSupport(0);
    }

    /**
     * Creates a new item set by copying another item set.
     *
     * @param itemSet The item set, which should be copied, as an instance of the class {@link
     *                ItemSet}. The item set may not be null
     */
    public ItemSet(@NotNull final ItemSet<ItemType> itemSet) {
        ensureNotNull(itemSet, "The item set may not be null");
        this.items = new TreeSet<>(itemSet.items);
        setSupport(itemSet.support);
    }

    /**
     * Returns the support of the item set.
     *
     * @return The support of the item set as a {@link Double} value. The support must be at least 0
     * and at maximum 1
     */
    public final double getSupport() {
        return support;
    }

    /**
     * Sets the support of the item set.
     *
     * @param support The support, which should be set, as a {@link Double} value. The support must
     *                be at least 0 and at maximum 1
     */
    public final void setSupport(final double support) {
        ensureAtLeast(support, 0, "The support must be at least 0");
        ensureAtMaximum(support, 1, "The support must be at least 1");
        this.support = support;
    }

    @Nullable
    @Override
    public final Comparator<? super ItemType> comparator() {
        return items.comparator();
    }

    @NotNull
    @Override
    public final SortedSet<ItemType> subSet(final ItemType fromElement, final ItemType toElement) {
        return items.subSet(fromElement, toElement);
    }

    @NotNull
    @Override
    public final SortedSet<ItemType> headSet(final ItemType toElement) {
        return items.headSet(toElement);
    }

    @NotNull
    @Override
    public final SortedSet<ItemType> tailSet(final ItemType fromElement) {
        return items.tailSet(fromElement);
    }

    @Override
    public final ItemType first() {
        return items.first();
    }

    @Override
    public final ItemType last() {
        return items.last();
    }

    @Override
    public final int size() {
        return items.size();
    }

    @Override
    public final boolean isEmpty() {
        return items.isEmpty();
    }

    @Override
    public final boolean contains(final Object o) {
        return items.contains(o);
    }

    @NotNull
    @Override
    public final Iterator<ItemType> iterator() {
        return items.iterator();
    }

    @NotNull
    @Override
    public final Object[] toArray() {
        return items.toArray();
    }

    @SuppressWarnings("SuspiciousToArrayCall")
    @NotNull
    @Override
    public final <T> T[] toArray(@NotNull final T[] a) {
        return items.toArray(a);
    }

    @Override
    public final boolean add(final ItemType item) {
        return items.add(item);
    }

    @Override
    public final boolean remove(final Object o) {
        return items.remove(o);
    }

    @Override
    public final boolean containsAll(@NotNull final Collection<?> c) {
        return items.containsAll(c);
    }

    @Override
    public final boolean addAll(@NotNull final Collection<? extends ItemType> c) {
        return items.addAll(c);
    }

    @Override
    public final boolean retainAll(@NotNull final Collection<?> c) {
        return items.retainAll(c);
    }

    @Override
    public final boolean removeAll(@NotNull final Collection<?> c) {
        return items.removeAll(c);
    }

    @Override
    public final void clear() {
        items.clear();
    }

    @Override
    public final int compareTo(@NotNull final ItemSet<ItemType> o) {
        return Double.compare(support, o.getSupport());
    }

    @Override
    public final ItemSet<ItemType> clone() {
        return new ItemSet<>(this);
    }

    @Override
    public final String toString() {
        return items.toString();
    }

    @Override
    public final int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + items.hashCode();
        return result;
    }

    @Override
    public final boolean equals(final Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        ItemSet<?> other = (ItemSet<?>) obj;
        return items.equals(other.items);
    }

}

File: src/main/java/de/mrapp/apriori/ItemSet.kt
/*
 * Copyright 2017 - 2019 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.apriori

import de.mrapp.util.Condition

import java.io.Serializable
import java.util.*

/**
 * An item set, which may contain several items. An item set may not contain duplicates. The
 * contained items are ordered according to the implementation of their [Comparable.compareTo]
 * method.
 *
 * @param ItemType The type of the items, which are contained by the item set
 * @author Michael Rapp
 * @since 1.0.0
 */
open class ItemSet<ItemType> : SortedSet<ItemType>, Comparable<ItemSet<ItemType>>, Serializable {

    /**
     * A sorted set, which contains the items, which are contained by the item set.
     */
    private val items: SortedSet<ItemType>

    /**
     * The support of the item set.
     */
    var support: Double = 0.0
        set(support) {
            Condition.ensureAtLeast(support, 0.0, "The support must be at least 0")
            Condition.ensureAtMaximum(support, 1.0, "The support must be at least 1")
            field = support
        }

    /**
     * Creates an empty item set.
     */
    constructor() {
        this.items = TreeSet()
        this.support = 0.0
    }

    /**
     * Creates a new item set by copying another [itemSet].
     */
    constructor(itemSet: ItemSet<ItemType>) {
        this.items = TreeSet(itemSet.items)
        this.support = itemSet.support
    }

    override val size: Int
        get() = items.size

    override fun comparator(): Comparator<in ItemType>? {
        return items.comparator()
    }

    override fun subSet(fromElement: ItemType, toElement: ItemType): SortedSet<ItemType> {
        return items.subSet(fromElement, toElement)
    }

    override fun headSet(toElement: ItemType): SortedSet<ItemType> {
        return items.headSet(toElement)
    }

    override fun tailSet(fromElement: ItemType): SortedSet<ItemType> {
        return items.tailSet(fromElement)
    }

    override fun first(): ItemType {
        return items.first()
    }

    override fun last(): ItemType {
        return items.last()
    }

    override fun isEmpty(): Boolean {
        return items.isEmpty()
    }

    override operator fun contains(element: ItemType): Boolean {
        return items.contains(element)
    }

    override fun iterator(): MutableIterator<ItemType> {
        return items.iterator()
    }

    override fun add(element: ItemType): Boolean {
        return items.add(element)
    }

    override fun remove(element: ItemType): Boolean {
        return items.remove(element)
    }

    override fun containsAll(elements: Collection<ItemType>): Boolean {
        return items.containsAll(elements)
    }

    override fun addAll(elements: Collection<ItemType>): Boolean {
        return items.addAll(elements)
    }

    override fun retainAll(elements: Collection<ItemType>): Boolean {
        return items.retainAll(elements)
    }

    override fun removeAll(elements: Collection<ItemType>): Boolean {
        return items.removeAll(elements)
    }

    override fun clear() {
        items.clear()
    }

    override fun compareTo(other: ItemSet<ItemType>) = support.compareTo(other.support)

    override fun toString() = items.toString()

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + items.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null)
            return false
        if (javaClass != other.javaClass)
            return false
        val another = other as ItemSet<*>?
        return items == another!!.items
    }

}
